lib/client.js

```diff
@@ -101,6 +101,7 @@ function Client (broker, conn, req) {
   }
 
   this.deliver0 = function deliverQoS0 (_packet, cb) {
+    _packet.topic = that.broker.persistence.restoreOriginalTopicFromSharedOne(_packet.topic)
```

Comment:
You could probably omit this call in case shared topics are not enabled or topic doesn't start with `$shared`?